examples/contrib/oed/gp_bayes_opt.py

```diff
@@ -0,0 +1,164 @@
+
+        :param function differentiable: a function amenable to torch
+            autograd
+        :param torch.Tensor x_init: the initial point
+
+        """
+        # transform x to an unconstrained domain
+        unconstrained_x_init = transform_to(self.constraints).inv(x_init)
+        unconstrained_x = torch.tensor(unconstrained_x_init, requires_grad=True)
+        minimizer = optim.LBFGS([unconstrained_x], max_eval=20)
+
+        def closure():
+            minimizer.zero_grad()
+            if (torch.log(torch.abs(unconstrained_x)) > 25.).any():
+                return torch.tensor(float('inf'))
+            x = transform_to(self.constraints)(unconstrained_x)
+            y = differentiable(x)
+            autograd.backward(unconstrained_x,
+                              autograd.grad(y, unconstrained_x, retain_graph=True))
+            return y
+
+        minimizer.step(closure)
+        # after finding a candidate in the unconstrained domain,
+        # convert it back to original domain.
+        x = transform_to(self.constraints)(unconstrained_x)
+        opt_y = differentiable(x)
+        return x.detach(), opt_y.detach()
+
+    def opt_differentiable(self, differentiable, num_candidates=5):
+        """Optimizes a differentiable function by choosing `num_candidates`
+        initial points at random and calling :func:`find_a_candidate` on
+        each. The best candidate is returned with its function value.
+
+        :param function differentiable: a function amenable to torch autograd
+        :param int num_candidates: the number of random starting points to
+            use
+        :return: the minimiser and its function value
+        :rtype: tuple
+        """
+
+        x_init = self.gpmodel.X[-1:].new_empty(1).uniform_(self.constraints.lower_bound,
+                                                           self.constraints.upper_bound)
+        candidates = []
+        values = []
+        for j in range(num_candidates):
+            x, y = self.find_a_candidate(differentiable, x_init)
+            candidates.append(x)
+            values.append(y)
+            x_init = x.new_empty(1).uniform_(self.constraints.lower_bound,
+                                             self.constraints.upper_bound)
+
+        mvalue, argmin = torch.min(torch.cat(values), dim=0)
+        return candidates[argmin.item()], mvalue
+
+    def acquire(self, method="Thompson", num_acquisitions=1, **opt_params):
+        """Selects `num_acquisitions` query points at which to query the
+        original function.
+
+        :param str method: the method to use for acquisition. Choose from:
+            Thompson
+        :param int num_acquisitions: the number of points to generate
+        :param dict opt_params: additional parameters for optimization
+            routines
+        :return: a tensor of points to evaluate `self.f` at
+        :rtype: torch.Tensor
+        """
+
+        if method == "Thompson":
```

Comment:
Let's split this out into an acquisition function parameter instead of implementing acquisition functions inside `acquire`, so it's easier to experiment with different Monte Carlo acquisition functions.